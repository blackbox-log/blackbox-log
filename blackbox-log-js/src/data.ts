import { Memoize as memoize } from 'typescript-memoize';

import { WasmPointer } from './wasm';

import type { Headers } from './headers';
import type { WasmExports, WasmObject } from './wasm';

export class DataParser implements WasmObject {
	readonly #wasm: WasmExports;
	readonly #ptr: WasmPointer;
	readonly #headers: Headers;

	constructor(wasm: WasmExports, headers: Headers, ptr: number) {
		this.#wasm = wasm;
		this.#headers = headers;
		this.#ptr = new WasmPointer(ptr, wasm.data_free);
	}

	free() {
		this.#ptr.free();
	}

	get isAlive(): boolean {
		return this.#ptr.isAlive;
	}

	get headers(): Headers {
		return this.#headers;
	}

	parseAll() {
		this.#wasm.data_parseAll(this.#ptr.ptr);
	}

	get mainFrameCount(): number {
		return this.#wasm.data_mainFrameCount(this.#ptr.ptr);
	}

	get gpsFrameCount(): number {
		return this.#wasm.data_gpsFrameCount(this.#ptr.ptr);
	}
}
