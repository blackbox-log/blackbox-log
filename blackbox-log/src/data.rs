//! Types for the data section of blackbox logs.

use alloc::vec::Vec;

use crate::event::{self, Event};
use crate::frame::{DataFrameKind, FrameKind, GpsFrame, GpsHomeFrame, MainFrame, SlowFrame};
use crate::parser::InternalError;
use crate::{Headers, Reader};

#[derive(Debug, Clone)]
pub(crate) struct Data {
    pub(crate) events: Vec<Event>,
    pub(crate) main_frames: Vec<FrameSync>,
    pub(crate) slow_frames: Vec<SlowFrame>,
    pub(crate) gps_frames: Vec<GpsFrame>,
    pub(crate) gps_home_frames: Vec<GpsHomeFrame>,
}

#[derive(Debug, Clone)]
pub(crate) struct FrameSync {
    pub(crate) main: MainFrame,
    pub(crate) slow: usize,
    pub(crate) gps: Option<usize>,
}

impl FrameSync {
    pub(crate) const fn new(main: MainFrame, slow: &[SlowFrame], gps: &[GpsFrame]) -> Self {
        Self {
            main,
            slow: slow.len() - 1,
            gps: gps.len().checked_sub(1),
        }
    }
}

/// Statistics about a decoded log.
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
#[non_exhaustive]
pub struct Stats {
    /// The number of valid frames found of each type.
    pub counts: FrameCounts,
}

#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct FrameCounts {
    pub main: usize,
    pub slow: usize,
    pub gps: usize,
    pub gps_home: usize,
}

impl Data {
    pub(crate) fn to_stats(&self) -> Stats {
        Stats {
            counts: FrameCounts {
                main: self.main_frames.len(),
                slow: self.slow_frames.len(),
                gps: self.gps_frames.len(),
                gps_home: self.gps_home_frames.len(),
            },
        }
    }

    pub(crate) fn parse(data: &mut Reader, headers: &Headers) -> Self {
        let mut events = Vec::new();
        let mut main_frames = Vec::new();
        let mut slow_frames = Vec::new();
        let mut gps_frames = Vec::new();
        let mut gps_home_frames = Vec::new();

        slow_frames.push(headers.slow_frames.empty_frame());

        if let Some(def) = &headers.gps_frames {
            gps_frames.push(def.empty_frame());
        }

        let mut restore;
        let mut last_kind = None;
        while let Some(byte) = data.read_u8() {
            restore = data.get_restore_point();

            let Some(kind) = FrameKind::from_byte(byte) else {
                tracing::debug!("found invalid frame byte: {byte:0>#2x}");

                if let Some(last_kind) = last_kind.take() {
                    data.restore(restore);

                    match last_kind {
                        FrameKind::Event => {
                            events.pop();
                        }
                        FrameKind::Data(DataFrameKind::Intra | DataFrameKind::Inter) => {
                            main_frames.pop();
                        }
                        FrameKind::Data(DataFrameKind::Slow) => {
                            slow_frames.pop();
                        }
                        FrameKind::Data(DataFrameKind::Gps) => {
                            gps_frames.pop();
                        }
                        FrameKind::Data(DataFrameKind::GpsHome) => {
                            gps_home_frames.pop();
                        }
                    };
                }

                skip_to_frame(data);
                continue;
            };

            tracing::trace!("trying to parse {kind:?} frame");

            let result = match kind {
                FrameKind::Event => match Event::parse_into(data, &mut events) {
                    Ok(event::EventKind::End) => {
                        tracing::trace!("found the end event");
                        break;
                    }
                    Ok(_) => Ok(()),
                    Err(err) => Err(err),
                },
                FrameKind::Data(DataFrameKind::Intra | DataFrameKind::Inter) => {
                    let frame = MainFrame::parse(data, kind, &main_frames, headers);
                    frame.map(|frame| {
                        main_frames.push(FrameSync::new(frame, &slow_frames, &gps_frames));
                    })
                }
                FrameKind::Data(DataFrameKind::Gps) => headers.gps_frames.as_ref().map_or_else(
                    || {
                        tracing::debug!("found GPS frame without GPS frame definition");
                        Err(InternalError::Retry)
                    },
                    |gps| {
                        gps.parse(
                            data,
                            headers,
                            main_frames.last().map(|sync| &sync.main),
                            gps_home_frames.last(),
                        )
                        .map(|frame| gps_frames.push(frame))
                    },
                ),
                FrameKind::Data(DataFrameKind::GpsHome) => {
                    headers.gps_home_frames.as_ref().map_or_else(
                        || {
                            tracing::debug!(
                                "found GPS home frame without GPS home frame definition"
                            );
                            Err(InternalError::Retry)
                        },
                        |gps_home| {
                            gps_home
                                .parse(data, headers)
                                .map(|frame| gps_home_frames.push(frame))
                        },
                    )
                }
                FrameKind::Data(DataFrameKind::Slow) => headers
                    .slow_frames
                    .parse(data, headers)
                    .map(|frame| slow_frames.push(frame)),
            };

            match result {
                Ok(()) => {
                    last_kind = Some(kind);
                }
                Err(InternalError::Retry) => {
                    tracing::debug!("found corrupted {kind:?} frame");
                    data.restore(restore);
                    skip_to_frame(data);
                }
                Err(InternalError::Eof) => {
                    tracing::debug!("found unexpected end of file in data section");
                    break;
                }
            }
        }

        Self {
            events,
            main_frames,
            slow_frames,
            gps_frames,
            gps_home_frames,
        }
    }
}

#[cold]
fn skip_to_frame(data: &mut Reader) {
    data.skip_until_any(
        &[
            FrameKind::Event,
            FrameKind::Data(DataFrameKind::Intra),
            FrameKind::Data(DataFrameKind::Slow),
            FrameKind::Data(DataFrameKind::Gps),
            FrameKind::Data(DataFrameKind::GpsHome),
        ]
        .map(u8::from),
    );
}
