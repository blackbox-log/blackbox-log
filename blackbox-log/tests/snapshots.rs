use std::collections::BTreeMap;
use std::fs;
use std::io::Read;

use blackbox_log::data::{ParseEvent, Stats};
use blackbox_log::event::Event;
use blackbox_log::units::{si, Flag, FlagSet};
use blackbox_log::{DataParser, Headers, Unit, Value};
use serde::ser::{SerializeStruct, Serializer};
use serde::Serialize;

macro_rules! run {
    () => {
        |path| {
            let mut file = fs::File::open(path).unwrap();
            let mut data = Vec::new();
            file.read_to_end(&mut data).unwrap();

            let file = blackbox_log::File::new(&data);
            let logs = file
                .iter()
                .map(|mut reader| {
                    Headers::parse(&mut reader).map(|headers| {
                        let data = DataParser::new(&mut reader, &headers);
                        LogSnapshot::new(&headers, data)
                    })
                })
                .collect::<Vec<_>>();

            insta::assert_ron_snapshot!(logs);
        }
    };
}

#[test]
fn own() {
    insta::glob!("logs/*.bbl", run!());
}

#[test]
#[ignore]
fn fc_blackbox() {
    insta::glob!("logs/fc-blackbox/*", run!());
}

#[test]
#[ignore]
fn gimbal_ghost() {
    insta::glob!("logs/gimbal-ghost/*", run!());
}

#[derive(Debug, Serialize)]
struct LogSnapshot<'data> {
    headers: Headers<'data>,
    stats: Stats,
    events: Vec<Event>,
    main: Fields,
    slow: Fields,
    gps: Fields,
}

impl<'data> LogSnapshot<'data> {
    fn new(headers: &Headers<'data>, mut data: DataParser<'data, '_, '_>) -> Self {
        let headers = headers.clone();

        let mut events = Vec::new();
        let mut main = headers.main_def().iter().collect::<Fields>();
        let mut slow = headers.slow_def().iter().collect::<Fields>();
        let mut gps = headers
            .gps_def()
            .iter()
            .flat_map(|def| def.iter())
            .collect::<Fields>();

        while let Some(frame) = data.next() {
            match frame {
                ParseEvent::Event(event) => events.push(event),
                ParseEvent::Main(frame) => main.update(frame),
                ParseEvent::Slow(frame) => slow.update(frame),
                ParseEvent::Gps(frame) => gps.update(frame),
            }
        }

        Self {
            headers,
            stats: data.stats().clone(),
            events,
            main,
            slow,
            gps,
        }
    }
}

#[derive(Debug, Serialize)]
struct Fields(Vec<FieldSnapshot>);

impl Fields {
    fn update<F: blackbox_log::frame::Frame>(&mut self, frame: F) {
        for (field, value) in self.0.iter_mut().zip(frame.iter()) {
            field.update(value.into());
        }
    }
}

impl<T, U> FromIterator<(T, U)> for Fields
where
    T: Into<String>,
    U: Into<Unit>,
{
    fn from_iter<I: IntoIterator<Item = (T, U)>>(iter: I) -> Self {
        Self(
            iter.into_iter()
                .map(|(name, unit)| FieldSnapshot::new(name.into(), unit.into()))
                .collect(),
        )
    }
}

#[derive(Debug, Serialize)]
struct FieldSnapshot {
    name: String,
    unit: Unit,
    history: History,
}

#[derive(Debug, Serialize)]
#[serde(untagged)]
enum History {
    Int(NumberHistory<16, i128>),
    Bool { yes: usize, no: usize },
    Flags(BTreeMap<&'static str, usize>),
}

#[derive(Debug)]
struct NumberHistory<const N: usize, T> {
    min: T,
    max: T,
    seen: Vec<T>,
    histogram: [usize; N],
}

impl FieldSnapshot {
    fn new(name: String, unit: Unit) -> Self {
        Self {
            name,
            unit,
            history: match unit {
                Unit::FrameTime
                | Unit::Amperage
                | Unit::Voltage
                | Unit::Acceleration
                | Unit::Rotation
                | Unit::GpsCoordinate
                | Unit::Altitude
                | Unit::Velocity
                | Unit::GpsHeading
                | Unit::Unitless => History::Int(NumberHistory::new()),
                Unit::Boolean => History::Bool { yes: 0, no: 0 },
                Unit::FlightMode | Unit::State | Unit::FailsafePhase => {
                    History::Flags(BTreeMap::new())
                }
            },
        }
    }

    #[allow(clippy::wildcard_enum_match_arm, clippy::cast_possible_truncation)]
    fn update(&mut self, value: Value) {
        match &mut self.history {
            History::Int(history) => history.update(match value {
                Value::FrameTime(t) => t.get::<si::time::microsecond>().round() as i128,
                Value::Amperage(a) => a.get::<si::electric_current::milliampere>().round() as i128,
                Value::Voltage(v) => v.get::<si::electric_potential::millivolt>().round() as i128,
                Value::Acceleration(a) => a
                    .get::<si::acceleration::centimeter_per_second_squared>()
                    .round() as i128,
                Value::Rotation(r) => {
                    r.get::<si::angular_velocity::degree_per_second>().round() as i128
                }
                Value::GpsCoordinate(c) => (c * 10000000.).round() as i128,
                Value::Altitude(a) => a.get::<si::length::meter>().round() as i128,
                Value::Velocity(v) => {
                    v.get::<si::velocity::centimeter_per_second>().round() as i128
                }
                Value::GpsHeading(h) => (h * 10.).round() as i128,
                Value::Unsigned(u) => u.into(),
                Value::Signed(s) => s.into(),
                _ => unreachable!(),
            }),
            History::Bool { yes, no } => match value {
                Value::Boolean(true) => *yes += 1,
                Value::Boolean(false) => *no += 1,
                _ => unreachable!(),
            },
            History::Flags(history) => {
                if let Value::FailsafePhase(phase) = value {
                    *history.entry(phase.as_name()).or_insert(0) += 1;
                } else {
                    let flags = match value {
                        Value::FlightMode(m) => m.as_names(),
                        Value::State(s) => s.as_names(),
                        _ => unreachable!(),
                    };

                    for flag in flags {
                        *history.entry(flag).or_insert(0) += 1;
                    }
                }
            }
        }
    }
}

impl<const N: usize, T> NumberHistory<N, T> {
    fn new() -> Self
    where
        T: Default,
    {
        Self {
            min: T::default(),
            max: T::default(),
            seen: Vec::new(),
            histogram: [0; N],
        }
    }

    fn update_range(&mut self, value: T)
    where
        T: PartialOrd,
    {
        if value < self.min {
            self.min = value;
        } else if value > self.max {
            self.max = value;
        }
    }

    /// Store new value sorted in `seen` if not already present
    fn update_seen(&mut self, value: T)
    where
        T: PartialOrd,
    {
        let index = self.seen.partition_point(|x| *x < value);
        if self.seen.get(index) != Some(&value) {
            self.seen.insert(index, value);
        }
    }

    /// Bucket so that sequential values go in sequential buckets, since data is
    /// usually clustered
    #[inline(always)]
    fn update_histogram(&mut self, value: u128) {
        self.histogram[(value % N as u128) as usize] += 1;
    }
}

impl<const N: usize> NumberHistory<N, i128> {
    fn update(&mut self, value: i128) {
        self.update_range(value);
        self.update_seen(value);
        self.update_histogram(value.unsigned_abs());
    }
}

impl<const N: usize> Serialize for NumberHistory<N, i128>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(self, self.min, self.max, serializer)
    }
}

impl<const N: usize> Serialize for NumberHistory<N, f64>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(
            self,
            format!("{:.2}", self.min),
            format!("{:.2}", self.max),
            serializer,
        )
    }
}

#[inline(always)]
fn serialize<const N: usize, T, S, U>(
    snapshot: &NumberHistory<N, T>,
    min: U,
    max: U,
    serializer: S,
) -> Result<S::Ok, S::Error>
where
    S: Serializer,
    U: Serialize,
    [usize; N]: Serialize,
{
    let unique = snapshot.seen.len();

    let mut state = serializer.serialize_struct("FieldSnapshot", 4)?;

    state.serialize_field("min", &min)?;
    state.serialize_field("max", &max)?;
    state.serialize_field("unique", &unique)?;

    let histogram = "histogram";
    if unique > 1 {
        state.serialize_field(histogram, &snapshot.histogram)?;
    } else {
        state.skip_field(histogram)?;
    }

    state.end()
}
