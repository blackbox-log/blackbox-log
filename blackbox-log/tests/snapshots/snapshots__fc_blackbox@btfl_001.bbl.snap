---
source: blackbox-log/tests/snapshots.rs
expression: logs
input_file: blackbox-log/tests/logs/fc-blackbox/btfl_001.bbl
---
[
  Ok(LogSnapshot(
    headers: Headers(
      version: V2,
      firmware_revision: "Betaflight 4.2.11 (948ba6339) STM32F7X2",
      firmware_kind: Betaflight,
      board_info: Some("ZEEZ ZEEZF7V2"),
      craft_name: None,
      vbat: Some(VbatConfig(
        reference: 2466,
        scale: 110,
      )),
      current_meter: Some(CurrentMeterConfig(
        offset: 0,
        scale: 250,
      )),
      acceleration_1g: Some(2048),
      gyro_scale: Some(1.0),
      min_throttle: Some(1070),
      motor_output_range: Some(MotorOutputRange(158, 2047)),
    ),
    stats: Stats(
      counts: FrameCounts(
        main: 98,
        slow: 3,
        gps_home: 0,
      ),
    ),
    events: [
      SyncBeep(32887122),
      FlightMode(
        flags: 0,
        last_flags: 1,
      ),
      Disarm(4),
      End,
    ],
    fields: Fields([
      FieldSnapshot(
        name: "loopIteration",
        min: 0,
        max: 1537,
        unique: 98,
        histogram: (7, 7, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6),
      ),
      FieldSnapshot(
        name: "time",
        min: 0,
        max: 33206271,
        unique: 98,
        histogram: (0, 1, 37, 5, 5, 5, 6, 1, 2, 3, 1, 11, 4, 0, 10, 7),
      ),
      FieldSnapshot(
        name: "axisP[0]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "axisP[1]",
        min: -2,
        max: 0,
        unique: 2,
        histogram: (97, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
      ),
      FieldSnapshot(
        name: "axisP[2]",
        min: 0,
        max: 1,
        unique: 2,
        histogram: (94, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
      ),
      FieldSnapshot(
        name: "axisI[0]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "axisI[1]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "axisI[2]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "axisD[0]",
        min: -2,
        max: 2,
        unique: 5,
        histogram: (80, 16, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
      ),
      FieldSnapshot(
        name: "axisD[1]",
        min: -4,
        max: 2,
        unique: 6,
        histogram: (91, 4, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
      ),
      FieldSnapshot(
        name: "axisF[0]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "axisF[1]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "axisF[2]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "rcCommand[0]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "rcCommand[1]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "rcCommand[2]",
        min: 0,
        max: 1,
        unique: 2,
        histogram: (37, 61, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
      ),
      FieldSnapshot(
        name: "rcCommand[3]",
        min: 0,
        max: 1000,
        unique: 1,
      ),
      FieldSnapshot(
        name: "setpoint[0]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "setpoint[1]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "setpoint[2]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "setpoint[3]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "vbatLatest",
        min: 0,
        max: 2463,
        unique: 4,
        histogram: (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 10, 39, 0, 19, 0, 30),
      ),
      FieldSnapshot(
        name: "amperageLatest",
        min: 0,
        max: 184,
        unique: 8,
        histogram: (70, 0, 0, 0, 8, 0, 0, 0, 10, 0, 0, 0, 10, 0, 0, 0),
      ),
      FieldSnapshot(
        name: "BaroAlt",
        min: 0,
        max: 279,
        unique: 7,
        histogram: (0, 12, 0, 0, 0, 12, 13, 6, 0, 0, 0, 0, 5, 25, 25, 0),
      ),
      FieldSnapshot(
        name: "rssi",
        min: 0,
        max: 784,
        unique: 2,
        histogram: (59, 0, 0, 0, 0, 0, 39, 0, 0, 0, 0, 0, 0, 0, 0, 0),
      ),
      FieldSnapshot(
        name: "gyroADC[0]",
        min: -1,
        max: 0,
        unique: 2,
        histogram: (94, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
      ),
      FieldSnapshot(
        name: "gyroADC[1]",
        min: 0,
        max: 1,
        unique: 2,
        histogram: (97, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
      ),
      FieldSnapshot(
        name: "gyroADC[2]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "accSmooth[0]",
        min: -56,
        max: 0,
        unique: 18,
        histogram: (3, 2, 2, 2, 4, 3, 5, 5, 2, 4, 7, 17, 17, 10, 8, 7),
      ),
      FieldSnapshot(
        name: "accSmooth[1]",
        min: -40,
        max: 0,
        unique: 26,
        histogram: (14, 3, 3, 3, 3, 3, 2, 4, 11, 5, 5, 4, 6, 8, 9, 15),
      ),
      FieldSnapshot(
        name: "accSmooth[2]",
        min: 0,
        max: 2061,
        unique: 14,
        histogram: (1, 8, 9, 5, 2, 4, 3, 14, 9, 9, 20, 6, 7, 1, 0, 0),
      ),
      FieldSnapshot(
        name: "motor[0]",
        min: 0,
        max: 172,
        unique: 16,
        histogram: (11, 10, 4, 5, 2, 0, 0, 1, 0, 2, 7, 1, 10, 15, 19, 11),
      ),
      FieldSnapshot(
        name: "motor[1]",
        min: 0,
        max: 183,
        unique: 16,
        histogram: (8, 16, 15, 6, 9, 10, 6, 4, 1, 2, 2, 0, 0, 1, 11, 7),
      ),
      FieldSnapshot(
        name: "motor[2]",
        min: 0,
        max: 179,
        unique: 17,
        histogram: (8, 10, 19, 7, 2, 6, 3, 3, 3, 1, 1, 0, 3, 4, 13, 15),
      ),
      FieldSnapshot(
        name: "motor[3]",
        min: 0,
        max: 183,
        unique: 15,
        histogram: (10, 12, 8, 9, 4, 4, 1, 4, 0, 1, 2, 0, 0, 16, 18, 9),
      ),
      FieldSnapshot(
        name: "flightModeFlags",
        min: 0,
        max: 1,
        unique: 2,
        histogram: (14, 84, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
      ),
      FieldSnapshot(
        name: "stateFlags",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "failsafePhase",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "rxSignalReceived",
        min: 0,
        max: 1,
        unique: 2,
        histogram: (1, 97, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
      ),
      FieldSnapshot(
        name: "rxFlightChannelsValid",
        min: 0,
        max: 1,
        unique: 2,
        histogram: (1, 97, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
      ),
    ]),
  )),
  Ok(LogSnapshot(
    headers: Headers(
      version: V2,
      firmware_revision: "Betaflight 4.2.11 (948ba6339) STM32F7X2",
      firmware_kind: Betaflight,
      board_info: Some("ZEEZ ZEEZF7V2"),
      craft_name: None,
      vbat: Some(VbatConfig(
        reference: 2439,
        scale: 110,
      )),
      current_meter: Some(CurrentMeterConfig(
        offset: 0,
        scale: 250,
      )),
      acceleration_1g: Some(2048),
      gyro_scale: Some(1.0),
      min_throttle: Some(1070),
      motor_output_range: Some(MotorOutputRange(158, 2047)),
    ),
    stats: Stats(
      counts: FrameCounts(
        main: 24893,
        slow: 9,
        gps_home: 0,
      ),
    ),
    events: [
      SyncBeep(191753984),
      FlightMode(
        flags: 0,
        last_flags: 1,
      ),
      Disarm(4),
      End,
    ],
    fields: Fields([
      FieldSnapshot(
        name: "loopIteration",
        min: 0,
        max: 398092,
        unique: 24893,
        histogram: (1556, 1556, 1556, 1556, 1556, 1556, 1556, 1556, 1556, 1556, 1556, 1556, 1556, 1555, 1555, 1555),
      ),
      FieldSnapshot(
        name: "time",
        min: 0,
        max: 241860649,
        unique: 24893,
        histogram: (1464, 1665, 1677, 1352, 1720, 945, 1801, 1696, 1465, 1765, 1794, 1336, 1670, 1020, 1814, 1709),
      ),
      FieldSnapshot(
        name: "axisP[0]",
        min: -41,
        max: 35,
        unique: 77,
        histogram: (2736, 2726, 2461, 2272, 2063, 1736, 1691, 1553, 1287, 1127, 1006, 1036, 906, 845, 768, 680),
      ),
      FieldSnapshot(
        name: "axisP[1]",
        min: -37,
        max: 40,
        unique: 78,
        histogram: (2777, 2740, 2379, 2220, 2092, 2038, 2011, 1805, 1516, 1224, 1054, 910, 704, 526, 457, 440),
      ),
      FieldSnapshot(
        name: "axisP[2]",
        min: -103,
        max: 79,
        unique: 183,
        histogram: (1923, 1971, 1960, 1641, 1555, 1515, 1572, 1603, 1530, 1405, 1418, 1410, 1244, 1414, 1406, 1326),
      ),
      FieldSnapshot(
        name: "axisI[0]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "axisI[1]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "axisI[2]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "axisD[0]",
        min: -15,
        max: 15,
        unique: 30,
        histogram: (10599, 7212, 3755, 1723, 736, 401, 207, 120, 62, 35, 23, 7, 4, 3, 1, 5),
      ),
      FieldSnapshot(
        name: "axisD[1]",
        min: -18,
        max: 11,
        unique: 29,
        histogram: (9945, 7163, 4060, 1915, 877, 441, 235, 139, 54, 34, 10, 7, 3, 5, 2, 3),
      ),
      FieldSnapshot(
        name: "axisF[0]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "axisF[1]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "axisF[2]",
        min: -1,
        max: 1,
        unique: 3,
        histogram: (24810, 83, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
      ),
      FieldSnapshot(
        name: "rcCommand[0]",
        min: 0,
        max: 1,
        unique: 2,
        histogram: (24320, 573, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
      ),
      FieldSnapshot(
        name: "rcCommand[1]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "rcCommand[2]",
        min: 0,
        max: 3,
        unique: 4,
        histogram: (9036, 15691, 102, 64, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
      ),
      FieldSnapshot(
        name: "rcCommand[3]",
        min: 0,
        max: 1000,
        unique: 1,
      ),
      FieldSnapshot(
        name: "setpoint[0]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "setpoint[1]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "setpoint[2]",
        min: 0,
        max: 1,
        unique: 2,
        histogram: (24757, 136, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
      ),
      FieldSnapshot(
        name: "setpoint[3]",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "vbatLatest",
        min: 0,
        max: 2507,
        unique: 42,
        histogram: (1445, 1567, 1977, 1328, 1795, 1748, 1289, 1477, 1666, 1080, 2428, 780, 1960, 1177, 1498, 1678),
      ),
      FieldSnapshot(
        name: "amperageLatest",
        min: 0,
        max: 948,
        unique: 158,
        histogram: (10529, 0, 0, 0, 5194, 0, 0, 0, 4465, 0, 0, 0, 4705, 0, 0, 0),
      ),
      FieldSnapshot(
        name: "BaroAlt",
        min: 0,
        max: 441,
        unique: 309,
        histogram: (1478, 1499, 1673, 1409, 2256, 2252, 1167, 1709, 2326, 1172, 946, 910, 951, 1662, 1727, 1756),
      ),
      FieldSnapshot(
        name: "rssi",
        min: 0,
        max: 914,
        unique: 107,
        histogram: (1794, 1193, 1315, 1294, 1392, 1293, 1494, 1987, 1991, 1590, 1694, 1490, 1292, 2391, 1096, 1587),
      ),
      FieldSnapshot(
        name: "gyroADC[0]",
        min: -22,
        max: 26,
        unique: 49,
        histogram: (2272, 3865, 3285, 2815, 2327, 1933, 1450, 1296, 1230, 1054, 850, 767, 549, 422, 383, 395),
      ),
      FieldSnapshot(
        name: "gyroADC[1]",
        min: -23,
        max: 21,
        unique: 45,
        histogram: (2602, 4321, 3455, 3425, 3115, 2350, 1546, 1095, 829, 592, 516, 497, 238, 159, 81, 72),
      ),
      FieldSnapshot(
        name: "gyroADC[2]",
        min: -46,
        max: 59,
        unique: 106,
        histogram: (1524, 2561, 2415, 2023, 2029, 1700, 1687, 1600, 1553, 1771, 1501, 865, 867, 1038, 956, 803),
      ),
      FieldSnapshot(
        name: "accSmooth[0]",
        min: -1058,
        max: 0,
        unique: 843,
        histogram: (1507, 1567, 1497, 1586, 1514, 1511, 1567, 1564, 1515, 1583, 1584, 1572, 1580, 1593, 1565, 1588),
      ),
      FieldSnapshot(
        name: "accSmooth[1]",
        min: -490,
        max: 321,
        unique: 795,
        histogram: (1562, 1565, 1596, 1586, 1623, 1567, 1609, 1575, 1569, 1584, 1596, 1585, 1514, 1424, 1470, 1468),
      ),
      FieldSnapshot(
        name: "accSmooth[2]",
        min: 0,
        max: 2166,
        unique: 503,
        histogram: (1537, 1570, 1561, 1472, 1455, 1471, 1564, 1586, 1504, 1573, 1582, 1546, 1623, 1645, 1615, 1589),
      ),
      FieldSnapshot(
        name: "motor[0]",
        min: 0,
        max: 604,
        unique: 432,
        histogram: (1362, 1401, 1431, 1362, 1374, 1438, 1370, 1316, 1397, 1329, 1309, 1436, 1622, 2241, 2934, 1571),
      ),
      FieldSnapshot(
        name: "motor[1]",
        min: 0,
        max: 531,
        unique: 331,
        histogram: (973, 901, 869, 856, 846, 841, 813, 872, 868, 922, 940, 1092, 3982, 8061, 1074, 983),
      ),
      FieldSnapshot(
        name: "motor[2]",
        min: 0,
        max: 515,
        unique: 335,
        histogram: (1379, 1384, 1368, 1318, 1231, 1228, 1307, 1257, 1285, 1322, 1245, 1381, 2476, 3718, 1482, 1512),
      ),
      FieldSnapshot(
        name: "motor[3]",
        min: 0,
        max: 681,
        unique: 505,
        histogram: (1321, 1267, 1289, 1251, 1208, 1178, 1207, 1134, 1137, 1230, 1178, 1217, 1964, 5422, 1508, 1382),
      ),
      FieldSnapshot(
        name: "flightModeFlags",
        min: 0,
        max: 1,
        unique: 2,
        histogram: (15, 24878, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
      ),
      FieldSnapshot(
        name: "stateFlags",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "failsafePhase",
        min: 0,
        max: 0,
        unique: 1,
      ),
      FieldSnapshot(
        name: "rxSignalReceived",
        min: 0,
        max: 1,
        unique: 2,
        histogram: (1, 24892, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
      ),
      FieldSnapshot(
        name: "rxFlightChannelsValid",
        min: 0,
        max: 1,
        unique: 2,
        histogram: (1, 24892, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
      ),
    ]),
  )),
]
